### PR TITLE
Wallet address resolution for Android

### PR DESCRIPTION
**Pull request by supermassive**
Resolves https://github.com/brave/brave-browser/issues/27464

## Submitter Checklist:

- [x] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run lint`, `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:

[untitled.webm](https://user-images.githubusercontent.com/168356/208616383-50fba0e7-adce-49f5-ad63-665292c9f663.webm)

### CODE REVIEW

File: android/java/org/chromium/chrome/browser/crypto_wallet/activities/BuySendSwapActivity.java [hunk 1]
```diff
@@ -1176,6 +1186,59 @@ public void onBackPressed() {
...
+                    onResolveWalletAddressDone(domain, response);
+                });
+                return true;
+            }
+        }
+
+        mResolvedAddrText.setVisibility(View.GONE);
+        mResolvedAddrText.setText("");
+        return false;
+    }
```
Comment (supermassive) [inline android/java/org/chromium/chrome/browser/crypto_wallet/activities/BuySendSwapActivity.java:1240]: Eventually will be refactored to a single mojo call https://github.com/brave/brave-browser/issues/26655

**Review by iefremov — APPROVED** [review]
browser/net LGTM

File: android/java/org/chromium/chrome/browser/crypto_wallet/activities/BuySendSwapActivity.java [hunk 2]
```diff
@@ -1176,6 +1186,59 @@ public void onBackPressed() {
...
 
+    private void onResolveWalletAddressDone(String domain, String result) {
+        if (!domain.equals(mSendToAddrText.getText().toString())) {
+            return;
+        }
+
+        if (result == null || result.isEmpty()) {
+            mResolvedAddrText.setVisibility(View.GONE);
+            mResolvedAddrText.setText("");
+            String notRegsteredErrorText = String.format(
```
Comment (SergeyZhukovsky) [inline android/java/org/chromium/chrome/browser/crypto_wallet/activities/BuySendSwapActivity.java:1197]: notRegsteredErrorText -> notRegisteredErrorText

File: android/java/org/chromium/chrome/browser/crypto_wallet/activities/BuySendSwapActivity.java [hunk 3]
```diff
@@ -1176,6 +1186,59 @@ public void onBackPressed() {
...
+        if (mCurrentBlockchainToken.coin == CoinType.ETH && WalletNativeUtils.isEnsTld(domain)) {
+            mJsonRpcService.ensGetEthAddr(domain, null,
+                    (response, requireOffchainConsent, errorResponse, errorString) -> {
+                        onResolveWalletAddressDone(domain, response);
+                    });
+            return true;
+        }
+
+        if (ChromeFeatureList.isEnabled(BraveFeatureList.BRAVE_WALLET_SNS)) {
+            if (mCurrentBlockchainToken.coin == CoinType.SOL
```
Comment (SergeyZhukovsky) [inline android/java/org/chromium/chrome/browser/crypto_wallet/activities/BuySendSwapActivity.java:1228]: could be combined to a single `if`

**Review by SergeyZhukovsky — APPROVED** [review]
++. There are two minor suggestions, but other than that looks good. Thanks for making it for Android.

File: browser/ipfs/content_browser_client_helper.cc [hunk 4]
```diff
@@ -80,16 +80,26 @@ bool HandleIPFSURLRewrite(GURL* url, content::BrowserContext* browser_context) {
...
+            g_browser_process->local_state())) {
+      return true;
+    }
+
+    if (decentralized_dns::IsUnstoppableDomainsTLD(url->host_piece()) &&
+        decentralized_dns::IsUnstoppableDomainsResolveMethodEnabled(
+            g_browser_process->local_state())) {
+      return true;
+    }
   }
```
Comment (yrliou) [inline browser/ipfs/content_browser_client_helper.cc:101]: Could we change ContentBrowserClientHelperUnitTest.HandleIPFSURLRewriteENS to ContentBrowserClientHelperUnitTest.HandleIPFSURLRewriteDDns and cover all cases there?
The logic is somewhat trivial but it's good to cover them all to ensure expected behavior and avoid regression in the future. Besides existing test cases, suggest to also test the case where IsLocalGatewayConfigured is false and IsXXXResolveMethodEnabled is true.

File: android/java/org/chromium/chrome/browser/crypto_wallet/activities/BuySendSwapActivity.java (hunk 3 above)
Comment (supermassive) [reply android/java/org/chromium/chrome/browser/crypto_wallet/activities/BuySendSwapActivity.java:1228]: This is supposed to be a temporary feature and to be removed as soon as our backend is ready. While inner if block matches pattern above

**Review by yrliou — APPROVED** [review]
LGTM

**Pull request merged**